language/libra-vm/src/libra_vm.rs

```diff
@@ -306,50 +321,66 @@ impl LibraVM {
+        };
+        let mut cost_strategy = CostStrategy::transaction(gas_schedule, txn_data.max_gas_amount());
+        let mut data_store = TransactionDataCache::new(remote_cache);
         // TODO: The logic for handling falied transaction fee is pretty ugly right now. Fix it later.
         let mut failed_gas_left = GasUnits::new(0);
         match payload {
             VerifiedTransactionPayload::Module(m) => {
-                self.move_vm.publish_module(m, &mut ctx, txn_data)
+                self.move_vm.publish_module(m, &mut data_store, txn_data)
             }
             VerifiedTransactionPayload::Script(s, ty_args, args) => {
-                let gas_schedule = match self.get_gas_schedule() {
-                    Ok(s) => s,
-                    Err(e) => return discard_error_output(e),
-                };
-                let ret =
-                    self.move_vm
-                        .execute_script(s, gas_schedule, &mut ctx, txn_data, ty_args, args);
-                let gas_usage = txn_data.max_gas_amount().sub(ctx.remaining_gas()).get();
+                let ret = self.move_vm.execute_script(
+                    s,
+                    ty_args,
+                    args,
+                    &mut cost_strategy,
+                    &mut data_store,
+                    txn_data,
+                );
+                let gas_usage = txn_data
+                    .max_gas_amount()
+                    .sub(cost_strategy.remaining_gas())
+                    .get();
                 TXN_EXECUTION_GAS_USAGE.observe(gas_usage as f64);
                 ret
             }
         }
         .map_err(|err| {
-            failed_gas_left = ctx.remaining_gas();
+            failed_gas_left = cost_strategy.remaining_gas();
             err
         })
         .and_then(|_| {
-            failed_gas_left = ctx.remaining_gas();
-            let mut gas_free_ctx = SystemExecutionContext::from(ctx);
-            self.run_epilogue(&mut gas_free_ctx, txn_data, account_currency_symbol)
-                .and_then(|_| {
-                    get_transaction_output(
-                        &mut gas_free_ctx,
-                        txn_data,
-                        VMStatus::new(StatusCode::EXECUTED),
-                    )
-                })
+            failed_gas_left = cost_strategy.remaining_gas();
+            let mut cost_strategy = CostStrategy::system(gas_schedule, failed_gas_left);
+            self.run_epilogue(
+                &mut data_store,
+                &mut cost_strategy,
+                txn_data,
+                account_currency_symbol,
+            )
+            .and_then(|_| {
+                get_transaction_output(
+                    &mut data_store,
+                    &cost_strategy,
+                    txn_data,
+                    VMStatus::new(StatusCode::EXECUTED),
+                )
+            })
         })
         .unwrap_or_else(|err| {
             self.failed_transaction_cleanup(
```

Comment:
This is a question for @sblackshear  and @tzakian . When a transaction fails should we just charge the claimed max gas amount sent by the user instead of the exact number of gas that was consumed? That would make this logic a lot cleaner. The difference is that for example when we hit an assertion failure in move, we will charge for full gas instead the gas used till the assertion failure.